packages/core/README.md

```diff
@@ -78,117 +62,378 @@ app.all("/mcp", async (c) => {
-  version: "1.0.0",
-  schemaAdapter: (schema) => z.toJSONSchema(schema as z.ZodType),
+const SearchSchema = z.object({
+  query: z.string(),
+  limit: z.number().optional(),
+  filters: z.array(z.string()).optional()
 });
+
+server.tool("search", {
+  inputSchema: SearchSchema,
+  handler: (args) => {
+    // args is typed as { query: string, limit?: number, filters?: string[] }
+    args.query.toLowerCase()
+    args.limit ?? 10
+    args.filters?.map(f => f.trim())
+
+    return { content: [{ type: "text", text: "..." }] }
+  }
+})
 ```
 
-#### Without schema adapter (JSON Schema only):
+### Structured Outputs
+
+Tools can return both human-readable content and machine-readable structured data. Use `outputSchema` to define the shape of `structuredContent`:
+
 ```typescript
-const server = new McpServer({
-  name: "my-server",
-  version: "1.0.0",
+const WeatherOutputSchema = z.object({
+  temperature: z.number(),
+  conditions: z.string(),
 });
+
+server.tool("getWeather", {
+  inputSchema: z.object({ location: z.string() }),
+  outputSchema: WeatherOutputSchema,
+  handler: (args) => ({
+    content: [{
+      type: "text",
+      text: `Weather in ${args.location}: 22°C, sunny`
+    }],
+    // structuredContent is typed and validated at runtime
+    structuredContent: {
+      temperature: 22,
+      conditions: "sunny",
+    }
+  })
+})
 ```
 
-## Connecting with Hono
+The `outputSchema` provides runtime validation and type inference for `structuredContent`.
+
+### Context API
+
+The handler context provides typed access to session data, authentication, and client capabilities:
 
 ```typescript
-import { Hono } from "hono";
-import { StreamableHttpTransport } from "mcp-lite";
+handler: (args, ctx) => {
+  ctx.progress?.({ progress: 50, total: 100 })
+  ctx.session?.id
+  ctx.authInfo?.userId
+  ctx.state.myCustomData = "..."
+
+  const validated = ctx.validate(MySchema, data)
+
+  if (ctx.client.supports("elicitation")) {
+    const result = await ctx.elicit({
+      message: "Confirm action?",
+      schema: z.object({ confirmed: z.boolean() })
+    })
+  }
+}
+```
 
-// Create transport and bind server
-const transport = new StreamableHttpTransport();
-const httpHandler = transport.bind(mcp);
+### Multiple Validation Libraries
 
-// Setup Hono app with MCP endpoint
-const app = new Hono();
-app.all("/mcp", async (c) => {
-  const response = await httpHandler(c.req.raw);
-  return response;
-});
+Use different validation libraries in the same server:
+
+```typescript
+import { z } from "zod"
+import * as v from "valibot"
+import { Schema } from "@effect/schema"
+
+server
+  .tool("zod-tool", { inputSchema: z.object({ ... }), handler: ... })
+  .tool("valibot-tool", { inputSchema: v.object({ ... }), handler: ... })
+  .tool("effect-tool", { inputSchema: Schema.struct({ ... }), handler: ... })
+  .tool("json-schema-tool", {
+    inputSchema: { type: "object", properties: { ... } },
+    handler: (args: MyType) => ...
+  })
```

Comment:
this example is wrong, right? we only support different validation libraries on different servers 

... but also makes me wonder: if you compose two servers with different validation libraries, what happens?